Hashes - clarify hash prefixes, SHA-512Half size.

mDuo13 opened the pull request:
- Fixes a typo where one old description of SHA-512Half said "bytes" when it meant "bits"
- Adds a table of hash prefixes based on https://github.com/ripple/rippled/blob/master/src/ripple/protocol/impl/HashPrefix.cpp but leaving off the types whose hashes I pretty much never expect to see.

mDuo13 committed:
Hashes - clarify hash prefixes, SHA-512Half size.

jwbusch commented on content/references/rippled-api/api-conventions/basic-data-types.md:
```diff
@@ -21,6 +21,24 @@ Each closed [Ledger](ledger-data-formats.html) has a [Ledger Index][] and a [Has
+| Transaction (Unsigned, for multi-signing)           | `0x534D5400`      | `SMT\0` |
+| Validation vote                                     | `0x56414C00`      | `VAL\0` |
+
+[Ledger objects IDs](ledger-object-ids.html) are calculated in a similar way, with 16-bit "space key" prefixes for each object type.
```
You use bytes to describe the size of prefixes before, but use bits here. What does "space key" mean? You might be able to omit the end of the sentence.

jwbusch commented on content/references/rippled-api/api-conventions/basic-data-types.md:
```diff
@@ -21,6 +21,24 @@ Each closed [Ledger](ledger-data-formats.html) has a [Ledger Index][] and a [Has
+
+In many cases, the XRP Ledger prefixes an object's binary data with a 4-byte code before calculating its hash, so that objects of different types have different hashes even if their binary formats are the same. The 4-byte codes are structured as three bytes of ASCII followed by a zero byte.
+
+The following table shows the hash prefixes for some common types of data:
```
Do we have all the types vs. just some? i.e. can we just fill out this table to cover all the cases?

mDuo13 replied:
I waffled on this a bit, trying to strike a balance between providing everything and not providing info that users will almost certainly not need.

In the end I kept finding more cases (like signing Claims for payment channels) where users might want to know a certain prefix I had thought they wouldn't, so I went with including everything.

jhaaaa reviewed (approved):
Looks good to me - thank you!

mDuo13 committed:
Revise hash prefixes table

mDuo13 committed:
Hash prefixes: re-add unsigned TX prefixes

These prefixes are relevant for implementing offline signing so they're
important to include. (The serialization docs want to know them.)

mDuo13 committed:
Hash prefixes: actually just re-add all prefixes

nbougalis commented on content/references/rippled-api/api-conventions/basic-data-types.md:
```diff
@@ -21,6 +21,33 @@ Each closed [Ledger](ledger-data-formats.html) has a [Ledger Index][] and a [Has
+### Hash Prefixes
+[[Source]<br>](https://github.com/ripple/rippled/blob/master/src/ripple/protocol/HashPrefix.h "Source")
+
+In many cases, the XRP Ledger prefixes an object's binary data with a 4-byte code before calculating its hash, so that objects of different types have different hashes even if their binary formats are the same. The 4-byte codes are structured as three bytes of ASCII followed by a zero byte.
```
>The 4-byte codes are structured as three bytes of ASCII followed by a zero byte.

Technically, this isn't required: the prefix is just four bytes long and, so far, we've used three ASCII characters and a zero byte, mostly because it simplfies debugging (you can just treat the thing as a string and dump it on the console or in a log file). But it could, just as well, be `0x4D454F57` (aka "MEOW").

Also note that "three bytes of ASCII" is a bit confusing. If you really want to say that, perhaps something like "current codes consist of 3 characters from the 26-character English alphabet (encoded as ASCII) followed by a 0 byte." You don't need to link to an ASCII table :smiley:

nbougalis commented on content/references/rippled-api/api-conventions/basic-data-types.md:
```diff
@@ -21,6 +21,33 @@ Each closed [Ledger](ledger-data-formats.html) has a [Ledger Index][] and a [Has
+| Unsigned Transaction (Single-signing) | N/A                                  | `0x53545800`      | `STX\0` |
+| Unsigned Transaction (Multi-signing)  | N/A                                  | `0x534D5400`      | `SMT\0` |
+| Validation vote                       | N/A                                  | `0x56414C00`      | `VAL\0` |
+| Validator ephemeral key update        | N/A                                  | `0x4D414E00`      | `MAN\0` |
```
Should we refer to this as a "Manifest"?

mDuo13 replied:
I mean, I thought we were going to call it a ["subkey authorization" per Confluence](https://ripplelabs.atlassian.net/wiki/spaces/~wilsonianb/pages/106004557/Renaming+Manifests) but I guess maybe we missed the train on that one?

nbougalis reviewed (commented):
Looks good. Left some minor comments.

mDuo13 committed:
Hash prefix clarifications per review
